src/storage/cache_proxy/cluster_manager.rs

```diff
@@ -0,0 +1,1096 @@
+        Ok(())
+    }
+
+    /// Update node info
+    pub async fn update_node_in_cluster(&self) -> DatenLordResult<()> {
+        let node = self.node.load();
+        let key = &KeyType::CacheNode(node.ip().to_owned());
+        let node_session = self.node_session.load();
+        let Some(current_session) = node_session.as_ref().cloned() else {
+            let current_node = self.node.load();
+            warn!(
+                "Current {} session is invalid, return to endpoint",
+                current_node.ip()
+            );
+            return Err(DatenLordError::CacheClusterErr {
+                context: vec![format!("Current {} session is invalid", current_node.ip())],
+            });
+        };
+
+        self.kv_engine
+            .set(
+                key,
+                &ValueType::Json(serde_json::to_value(node.as_ref())?),
+                Some(SetOption {
+                    session: Some(current_session),
+                    prev_kv: false,
+                }),
+            )
+            .await
+            .with_context(|| "Failed to register node to etcd".to_owned())?;
+
+        Ok(())
+    }
+
+    /// Register current node to etcd and keep alive
+    pub async fn register(&self) -> DatenLordResult<()> {
+        // Get current node info
+        let current_node_info = self.node.load();
+        debug!("register: {} to etcd", current_node_info.ip());
+        let node_session = self.node_session.load();
+        let Some(current_session) = node_session.as_ref().cloned() else {
+            let current_node = self.node.load();
+            warn!(
+                "Current {} session is invalid, return to endpoint",
+                current_node.ip()
+            );
+            return Err(DatenLordError::CacheClusterErr {
+                context: vec![format!("Current {} session is invalid", current_node.ip())],
+            });
+        };
+
+        // Try to register current node to etcd
+        self.kv_engine
+            .set(
+                &KeyType::CacheNode(current_node_info.ip().to_owned()),
+                &ValueType::Json(serde_json::to_value(current_node_info.as_ref().clone())?),
+                Some(SetOption {
+                    // Set lease
+                    session: Some(current_session),
+                    prev_kv: false,
+                }),
+            )
+            .await
+            .with_context(|| "Failed to register node to etcd".to_owned())?;
+
+        debug!("register: {} to etcd success", current_node_info.ip());
+
+        Ok(())
+    }
+
+    /// Do campaign
+    ///
+    /// Try to campaign master, will return status and master value
+    pub async fn do_campaign(&self) -> DatenLordResult<(bool, String)> {
+        let node = self.get_node();
+        let hash_ring = self.get_ring(false).await?;
+
+        let node_session = self.node_session.load();
+        let session = if let Some(session) = node_session.as_ref().cloned() {
```

Comment:
the code is long and repeat many times  (session.load & validation & warnning log),  and repeat many times ,  should be put in the function